elasticdl/doc/async_sgd_design.md

```diff
@@ -0,0 +1,118 @@
+# Design Doc: Asynchronous SGD
+
+## Motivation
+Parameter server (PS) based distributed training uses data parallelism to speed up training. 
+
+We have implemented synchronous SGD in ElasticDL. When PS accumulates `grads_to_wait` gradients from workers, PS averages these gradients and updates the model with the averaged gradients. PS also maintains `model_version` which equals to the number of model updates. Each worker has a local copy of the model. Before a minibatch training step starts, if there is a new `model_version` on PS, the worker will get the new model from PS to replace the local model. After computing the gradients with a minibatch, the worker reports the gradients to PS together with the local model version. When PS receives gradients from a worker, it only accepts the gradients with a model version same as the current PS `model_version`.
+
+
+
+This Synchronous SGD ensures model consistency in the price of wasted and blocked computation. 
+
+* Wasted computation: when a worker reports gradients with an outdated model version to PS, PS will reject these gradients. The worker have to get the current model from PS, reuse the minibatch data to train the model again.
+* Blocked computation: PS has to use a lock for model update with gradients and model read by workers to ensure model consistency.
+
```

Comment:
I think that the main problem of BSP is that a global barrier is needed to ensure PS get gradients from all workers. The fast worker has to wait for slow workers.

Reply:
That is the BSP implemented by TensorFlow. In each iteration, if PS needs *n* gradients to accumulate, it places *n* tokens in a queue. When a worker takes a token from the queue, it starts minibatch training and reports gradients to PS. When PS gets *n* gradients, it updates the model, and places another *n* tokens in the queue. If there are no tokens in the queue, workers need to wait for new tokens.
In ElasticDL, we have implemented a task-based synchronous SGD. Workers do not wait after `ReportGradient`, but continue to train with the next minibatch. When the worker reports gradients to PS, PS may already have updated the model and rejects the gradients. This minibatch training is wasted.